File: Binary_Search/034.Find-First-And-Last-Position-of-Element-in-Sorted-Array/034.Find-First-And-Last-Position-of-Element-in-Sorted-Array.ts
function searchRange(nums: number[], target: number): number[] {
  const result = [-1, -1];

  let left = 0;
  let right = nums.length - 1;
  while (left < right) {
    const middle = left + Math.floor((right - left) / 2);
    if (nums[middle] >= target) {
      right = middle;
    } else {
      left = middle + 1;
    }
  }
  if (nums[right] === target) {
    result[0] = right;
    while (right + 1 < nums.length && nums[right + 1] === target) {
      right += 1;
    }
    result[1] = right;
  }
  return result;
}

File: Binary_Search/034.Find-First-And-Last-Position-of-Element-in-Sorted-Array/034.Find-First-And-Last-Position-of-Element-in-Sorted-Array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if len(nums) == 0:
            return [-1, -1]

        left = 0
        right = len(nums) - 1
        while left < right:
            middle = (left + right) // 2
            if nums[middle] >= target:
                right = middle
            else:
                left = middle + 1
        
        startIndex = right if nums[right] == target else -1
        
        left = 0
        right = len(nums) - 1
        while left < right:
            middle = math.ceil((left + right) / 2)
            if nums[middle] <= target:
                left = middle
            else:
                right = middle - 1
        
        endIndex = left if nums[left] == target else -1
        return [startIndex, endIndex]
